### PR TITLE
log-backup: Using PD as metastore

### PR DESCRIPTION
**Pull request by YuJuncen**
### What is changed and how it works?

Issue Number: Close https://github.com/tikv/tikv/issues/13867

What's Changed:
This PR implemented a PD based metastore, and make it the default for log backup.
- I added a new trait `MetaStorageClient` to the package `pd_client`, which is the API set for interacting with a MetaStorage service.
- This trait has been implemented for `client::RpcClient`, because `MetaStorage` is another service rather than the `PD` service, I reused the channel we have dialed to PD. 
- There are some `MetaStorageClient` wrappers for general purpose, `AutoHeader` for adding request headers, `Checked` for checking the response.  
- There is a new method `get_latest` in the `MetaStorage` trait, when we don't need snapshot read, this interface can reduce a RTT.

```commit-message
This also makes `etcd-client` and `tonic` optional requirements, you can enable them by `metastore-etcd`.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by YuJuncen** [comment]
Hi @nolouch please note that the current implementation is based on the `GlobalConfig` API and may be hugely changed in the future, perhaps you don't need to review this PR for now.

**Comment by YuJuncen** [comment]
@nolouch The PR is ready for review now.
also cc @BusyJay and @tabokie , please help to check the `pd_client` part (only implemented for v1 now, do I need to implement `MetaStorageClient` for v2 now?), thanks.

**Comment by HuSharp** [comment]
Maybe we can remove the `GlobalConfig` interface in another pr?

File: components/pd_client/src/meta_storage.rs [hunk 1]
```diff
@@ -0,0 +1,302 @@
+// Copyrighfield1 2023 TiKV Project Authors. Licensed under Apache-2.0.
```
Comment (rleungx) [inline components/pd_client/src/meta_storage.rs:1]:
```suggestion
// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
```

**Review by Leavrth — APPROVED** [review]

File: components/backup-stream/src/endpoint.rs [hunk 2]
```diff
@@ -619,7 +619,6 @@ where
             let task_clone = task.clone();
             let run = async move {
                 let task_name = task.info.get_name();
-                cli.init_task(&task.info).await?;
```
Comment (bufferflies) [inline components/backup-stream/src/endpoint.rs:622]: why remove this?

Comment (YuJuncen) [reply components/backup-stream/src/endpoint.rs:622]: This is a legacy code which initializes some keys not being used any more.

Comment (YuJuncen) [reply components/backup-stream/src/endpoint.rs:622]: (Also this relies on the `Txn` API which isn't supported by the PD API.)

File: components/server/src/server.rs [hunk 3]
```diff
@@ -1039,17 +1037,13 @@ where
...
-                    tls: Arc::clone(&self.security_mgr),
-                },
-            );
             let backup_stream_endpoint = backup_stream::Endpoint::new(
                 node.id(),
-                etcd_cli,
+                PdStore::new(Checked::new(AutoHeader::new(
+                    Arc::clone(&self.pd_client),
+                    pd_client::Source::LogBackup,
+                    self.pd_client.get_cluster_id().unwrap_or_default(),
```
Comment (bufferflies) [inline components/server/src/server.rs:1045]: how about remove this parameter of cluster id?

Comment (YuJuncen) [reply components/server/src/server.rs]: Done

File: components/pd_client/src/meta_storage.rs [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+        Self { inner, source }
+    }
+
+    fn prepare_header(&self, h: &mut pb::RequestHeader) {
+        h.set_source(self.source.to_string());
+    }
+}
+
+impl<S: MetaStorageClient> MetaStorageClient for Sourced<S> {
+    type WatchStream = <S as MetaStorageClient>::WatchStream;
```
Comment (3pointer) [inline components/pd_client/src/meta_storage.rs:172]:
```suggestion
    type WatchStream = S::WatchStream;
```

File: components/pd_client/src/meta_storage.rs [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+            .boxed()
+    }
+
+    fn watch(&self, req: Watch) -> Self::WatchStream {
+        CheckedStream(self.0.watch(req))
+    }
+}
+
+impl<S: MetaStorageClient> MetaStorageClient for Arc<S> {
+    type WatchStream = <S as MetaStorageClient>::WatchStream;
```
Comment (3pointer) [inline components/pd_client/src/meta_storage.rs:275]:
```suggestion
    type WatchStream = S::WatchStream;
```

Comment (YuJuncen) [reply components/pd_client/src/meta_storage.rs]: done

File: components/pd_client/src/meta_storage.rs (hunk 4 above)
Comment (YuJuncen) [reply components/pd_client/src/meta_storage.rs]: done

File: components/pd_client/src/client.rs [hunk 6]
```diff
@@ -1117,3 +1122,98 @@ impl PdClient for RpcClient {
...
+        let timer = Instant::now();
+        self.fill_cluster_id_for(req.inner.mut_header());
+        let executor = move |client: &Client, req: GetRequest| {
+            let handler = {
+                let inner = client.inner.rl();
+                inner
+                    .meta_storage
+                    .get_async_opt(&req, call_option_inner(&inner))
+                    .unwrap_or_else(|e| {
+                        panic!("fail to request PD {} err {:?}", "metastorage::get", e)
```
[the remark this answers is not in the document]
Comment (YuJuncen) [reply components/pd_client/src/client.rs:1143]: (In fact this error can be removed by mapping the error into `PdError::Grpc(...)`, should we do so?)

**Review by bufferflies — APPROVED** [review]

**Review by 3pointer — APPROVED** [review]
LGTM

**Review by hicqu — APPROVED** [review]

**Pull request merged**